### DISCUSSION TITLE
Trigger after login?

### DISCUSSION THREAD

**Discussion by howesteve** [supabase/supabase#7463] [Questions] [answered]
Hello,

Is there a way to trigger a server function after a user is logged in? 
Specifically, I want to keep track of a profiles.last_seen field with the timestamp of the latest user's login.

Thanks,
Howe

**Comment by silentworks** [comment, marked as answer]
There is a `last_signed_in_at` column on the `auth.users` table, you could create a trigger to fire when a user signs in.
```sql
create function public.create_last_signed_in_on_profiles()
    returns trigger as $$
    begin
      IF (NEW.last_sign_in_at is null) THEN
        RETURN NULL;
      ELSE
        UPDATE public.profiles
        SET last_signed_in = NEW.last_sign_in_at
        WHERE id = (NEW.id)::uuid;
        RETURN NEW;
      END IF;
    end;
    $$ language plpgsql security definer;
```
And create an after update trigger

```sql
create trigger on_last_signed_in
    after update on auth.users
    for each row execute procedure public.create_last_signed_in_on_profiles();
```
Do note I haven't tested this but it should work.

**Reply by howesteve** [reply]
Thanks, I had missed it. It works.

**Comment by DanMossa** [comment]
@silentworks And where/how does `auth.last_sign_in_at` get populated?

**Reply by mb21** [reply]
@silentworks We have the same question: when is `auth.last_sign_in_at` updated? We have users that show their `last_sign_in_at` from two weeks ago, yet they confirm they logged in just yesterday and the "JWT expiry limit" is set to 1 hour.

edit: hm, [the code](https://github.com/supabase/gotrue/blob/206fc0908992e6c22a6343a7a7517f66322764b5/internal/models/refresh_token.go#L143) seems to update `last_sign_in_at` whenever a new refresh token is created, but not when a new access_token is created. So it shows when a user last time entered their password (or similar), that is a bit unexpected to me... any way to look up when a user got a new refresh token?

(note for those who wonder: "Refresh tokens are one-time use and never expire.", from the authentication settings page)
